File: app/src/main/java/tech/diarmaid/koohiiaite/adapter/KanjiDetailAdapter.java
package tech.diarmaid.koohiiaite.adapter;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import tech.diarmaid.koohiiaite.database.dao.HeisigKanjiDataSource;
import tech.diarmaid.koohiiaite.database.dao.KeywordDataSource;
import tech.diarmaid.koohiiaite.database.dao.UserKeywordDataSource;
import tech.diarmaid.koohiiaite.fragment.DictionaryFragment;
import tech.diarmaid.koohiiaite.fragment.KoohiiFragment;
import tech.diarmaid.koohiiaite.fragment.SampleWordsFragment;
import tech.diarmaid.koohiiaite.fragment.StoryFragment;
import tech.diarmaid.koohiiaite.model.HeisigKanji;
import tech.diarmaid.koohiiaite.model.Keyword;

import static tech.diarmaid.koohiiaite.utils.Constants.KANJI_DETAIL_TABS;

/**
 * Invoked when user clicks a list item in the KanjiListActivity
 */
public class KanjiDetailAdapter extends FragmentPagerAdapter {
    private StoryFragment storyFragment;
    private DictionaryFragment dictionaryFragment;
    private SampleWordsFragment sampleWordsFragment;
    private KoohiiFragment koohiiFragment;

    private Context mContext;
    private Bundle arguments;

    public KanjiDetailAdapter(FragmentManager fragmentManager, Bundle arguments, Context context) {
        super(fragmentManager);
        this.mContext = context;
        int heisigId = arguments.getInt("heisigId");
        String keyword = getKeywordFromDatabase(heisigId);
        String userKeyword = getUserKeywordFromDatabase(heisigId);
        String kanji = getKanjiFromDatabase(heisigId);
        arguments.putString("keyword", keyword);
        arguments.putString("userKeyword", userKeyword);
        arguments.putString("kanji", kanji);
        this.arguments = arguments;
    }

    private String getKanjiFromDatabase(int heisigId)
    {
        HeisigKanjiDataSource dataSource = new HeisigKanjiDataSource(mContext);
        dataSource.open();
        HeisigKanji heisigKanji = dataSource.getKanjiFor(heisigId);
        dataSource.close();

        return heisigKanji.getKanji();
    }

    private String getKeywordFromDatabase(int heisigId)
    {
        KeywordDataSource dataSource = new KeywordDataSource(mContext);
        dataSource.open();
        Keyword keyword = dataSource.getKeywordFor(heisigId);
        dataSource.close();

        return keyword.getKeywordText();
    }

    private String getUserKeywordFromDatabase(int heisigId)
    {
        UserKeywordDataSource dataSource = new UserKeywordDataSource(mContext);
        dataSource.open();
        Keyword keyword = dataSource.getKeywordFor(heisigId);
        dataSource.close();

        return keyword == null ? null : keyword.getKeywordText();
    }

    // Returns total number of pages
    @Override
    public int getCount() {
        return KANJI_DETAIL_TABS;
    }

    // Returns the fragment to display for that page
    @Override
    public Fragment getItem(int position) {
        switch (position) {
            case 0:
                if(storyFragment == null) {
                    storyFragment = new StoryFragment();
                    storyFragment.setArguments(arguments);
                }
                return storyFragment;
            case 1:
                if(dictionaryFragment == null) {
                    dictionaryFragment = new DictionaryFragment();
                    dictionaryFragment.setArguments(arguments);
                }
                return dictionaryFragment;
            case 2:
                if(sampleWordsFragment == null) {
                    sampleWordsFragment = new SampleWordsFragment();
                    sampleWordsFragment.setArguments(arguments);
                }
                return sampleWordsFragment;
            case 3:
                if(koohiiFragment == null) {
                    koohiiFragment = new KoohiiFragment();
                    koohiiFragment.setArguments(arguments);
                }
                return koohiiFragment;
            default:
                return null;
        }
    }

    // Returns the page title for the top indicator
    @Override
    public CharSequence getPageTitle(int position) {
        switch (position) {
            case 0:
                return "Story";
            case 1:
                return "Dictionary";
            case 2:
                return "Sample Words";
            case 3:
                return "Koohii";
            default:
                return "Undefined";
        }
    }

}

File: app/src/main/java/tech/diarmaid/koohiiaite/adapter/KanjiDetailAdapter.kt
package tech.diarmaid.koohiiaite.adapter

import android.content.Context
import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import tech.diarmaid.koohiiaite.database.AppDatabase
import tech.diarmaid.koohiiaite.fragment.DictionaryFragment
import tech.diarmaid.koohiiaite.fragment.KoohiiFragment
import tech.diarmaid.koohiiaite.fragment.SampleWordsFragment
import tech.diarmaid.koohiiaite.fragment.StoryFragment
import tech.diarmaid.koohiiaite.utils.Constants.KANJI_DETAIL_TABS
import tech.diarmaid.koohiiaite.viewmodel.KanjiDetailViewModel
import kotlin.coroutines.CoroutineContext

/**
 * Invoked when user clicks a list item in the KanjiListActivity
 */
class KanjiDetailAdapter(fragmentManager: FragmentManager, private val arguments: Bundle, private val mContext: Context, viewModel: KanjiDetailViewModel?, viewLifecycleOwner: LifecycleOwner) : FragmentPagerAdapter(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT), CoroutineScope {
    private var storyFragment: StoryFragment? = null
    private var dictionaryFragment: DictionaryFragment? = null
    private var sampleWordsFragment: SampleWordsFragment? = null
    private var koohiiFragment: KoohiiFragment? = null
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Default

    init {
        viewModel?.heisigId?.observe(viewLifecycleOwner, Observer { heisigId ->
            if (heisigId > 0) {
                launch(Dispatchers.IO) {
                    val kanji = AppDatabase.getDatabase(mContext).heisigKanjiDao().getKanjiFor(heisigId)?.kanji
                    val keyword = AppDatabase.getDatabase(mContext).keywordDao().getKeywordFor(heisigId)?.keywordText
                    val userKeyword = AppDatabase.getDatabase(mContext).userKeywordDao().getKeywordFor(heisigId)?.keywordText
                            ?: ""
                    Log.d("KanjiDetailAdapter", "Async id: $heisigId, kanji $kanji, keyword : $keyword, userkeyword : $userKeyword")
                    launch(Dispatchers.Main) {
                        viewModel.kanji.value = kanji
                        viewModel.keyword.value = keyword
                        viewModel.userKeyword.value = userKeyword
                        Log.d("KanjiDetailAdapter", "Posted id: $heisigId, kanji $kanji, keyword : $keyword, userkeyword : $userKeyword")
                    }
                }
            }
        })
    }

    private fun getKanjiFromDatabase(heisigId: Int): LiveData<String> {
        val data = MutableLiveData<String>()
        launch(Dispatchers.IO) {
            val dataSource = AppDatabase.getDatabase(mContext).heisigKanjiDao()
            data.postValue(dataSource.getKanjiFor(heisigId)?.kanji)
        }
        return data
    }

    private fun getKeywordFromDatabase(heisigId: Int): LiveData<String> {
        val data = MutableLiveData<String>()
        launch(Dispatchers.IO) {
            val dataSource = AppDatabase.getDatabase(mContext).keywordDao()
            data.postValue(dataSource.getKeywordFor(heisigId)?.keywordText)
        }
        return data
    }

    private fun getUserKeywordFromDatabase(heisigId: Int): LiveData<String?> {
        val data = MutableLiveData<String>()
        launch(Dispatchers.IO) {
            val dataSource = AppDatabase.getDatabase(mContext).userKeywordDao()
            data.postValue(dataSource.getKeywordFor(heisigId)?.keywordText)
        }
        return data
    }

    // Returns total number of pages
    override fun getCount(): Int {
        return KANJI_DETAIL_TABS
    }

    // Returns the fragment to display for that page
    override fun getItem(position: Int): Fragment {
        when (position) {
            0 -> {
                if (storyFragment == null) {
                    storyFragment = StoryFragment()
                    storyFragment?.arguments = arguments
                }
                return storyFragment as StoryFragment
            }
            1 -> {
                if (dictionaryFragment == null) {
                    dictionaryFragment = DictionaryFragment()
                    dictionaryFragment?.arguments = arguments
                }
                return dictionaryFragment as DictionaryFragment
            }
            2 -> {
                if (sampleWordsFragment == null) {
                    sampleWordsFragment = SampleWordsFragment()
                    sampleWordsFragment?.arguments = arguments
                }
                return sampleWordsFragment as SampleWordsFragment
            }
            3 -> {
                if (koohiiFragment == null) {
                    koohiiFragment = KoohiiFragment()
                    koohiiFragment?.arguments = arguments
                }
                return koohiiFragment as KoohiiFragment
            }
            else -> return getItem(0)
        }
    }

    // Returns the page title for the top indicator
    override fun getPageTitle(position: Int): CharSequence? {
        return when (position) {
            0 -> "Story"
            1 -> "Dictionary"
            2 -> "Sample Words"
            3 -> "Koohii"
            else -> "Undefined"
        }
    }
}
